File: Library/DSA Reference/Data Structures/5) Stacks/stacks.cpp
#include <bits/stdc++.h>
using namespace std;

/*
    ============================================================
    STACKS CHEAT SHEET — C++20
    ============================================================

    CORE / MUST KNOW
    ----------------
    Main DSA type: stack<T>

    Stack = LIFO
        Last In, First Out

    push / push_back      O(1) amortized in common back-based forms
    pop                   O(1)
    top                   O(1)
    size                  O(1)
    empty                 O(1)

    std::stack is a container adaptor.
    It does NOT provide iterators or random indexing.

    EXTRAS / REFERENCE
    ------------------
    emplace, swap, custom underlying container, vector/list/deque
    implementations, safe printing via copying, manual array/linked-
    list stacks, C++23 push_range, etc.
*/

int main() {

    // ============================================================
    // CORE — MUST KNOW
    // ============================================================

    // ------------------------------------------------------------
    // 1. INITIALISATION
    // ------------------------------------------------------------

    stack<int> s;

    // ------------------------------------------------------------
    // 2. PUSH
    // ------------------------------------------------------------

    s.push(32);
    s.push(10);
    s.push(20);

    // Stack:
    // top -> 20
    //        10
    //        32

    // ------------------------------------------------------------
    // 3. TOP
    // ------------------------------------------------------------

    if (!s.empty()) {
        int top = s.top();              // O(1)
        cout << "Top: " << top << '\n';
    }

    // IMPORTANT:
    // Calling top() on an empty stack is invalid.
    // Always check empty() first when emptiness is possible.

    // ------------------------------------------------------------
    // 4. POP
    // ------------------------------------------------------------

    if (!s.empty())
        s.pop();                        // O(1)

    // pop() removes the top element.
    // It does NOT return the removed value.

    // If you need the value:
    if (!s.empty()) {
        int x = s.top();
        s.pop();
        cout << "Popped: " << x << '\n';
    }

    // ------------------------------------------------------------
    // 5. SIZE
    // ------------------------------------------------------------

    int size = (int)s.size();           // O(1)
    cout << "Size: " << size << '\n';

    // ------------------------------------------------------------
    // 6. EMPTY
    // ------------------------------------------------------------

    bool isEmpty = s.empty();            // O(1)
    cout << "Empty: " << boolalpha << isEmpty << '\n';

    // ------------------------------------------------------------
    // 7. COMMON STACK PATTERN
    // ------------------------------------------------------------

    while (!s.empty()) {
        cout << s.top() << ' ';
        s.pop();
    }
    cout << '\n';

    /*
        This prints from TOP -> BOTTOM and destroys the stack.

        TC: O(n)
        SC: O(1) auxiliary
    */

    // ------------------------------------------------------------
    // 8. REBUILD + SAFE COPY FOR PRINTING
    // ------------------------------------------------------------

    stack<int> original;

    original.push(1);
    original.push(2);
    original.push(3);
    original.push(4);

    stack<int> copy = original;          // O(n)

    while (!copy.empty()) {
        cout << copy.top() << ' ';
        copy.pop();
    }

    cout << '\n';

    /*
        original is preserved.
        TC: O(n)
        SC: O(n) because of the copied stack
    */

    // ============================================================
    // EXTRAS — REFERENCE
    // ============================================================

    // ------------------------------------------------------------
    // 9. EMPLACE
    // ------------------------------------------------------------

    stack<pair<int, string>> pairs;

    pairs.emplace(1, "one");             // Construct element in place

    if (!pairs.empty()) {
        cout << pairs.top().first
             << ' '
             << pairs.top().second << '\n';
    }

    // ------------------------------------------------------------
    // 10. SWAP
    // ------------------------------------------------------------

    stack<int> a;
    stack<int> b;

    a.push(1);
    b.push(2);

    a.swap(b);                           // O(1) for compatible underlying
                                         // containers

    // ------------------------------------------------------------
    // 11. CUSTOM UNDERLYING CONTAINER
    // ------------------------------------------------------------

    stack<int, deque<int>> dqStack;      // default-style choice

    stack<int, vector<int>> vecStack;    // vector-backed stack

    stack<int, list<int>> listStack;     // list-backed stack

    dqStack.push(1);
    vecStack.push(2);
    listStack.push(3);

    /*
        std::stack is an adaptor.

        Common underlying containers:
            deque  -> default
            vector
            list

        For normal DSA, `stack<int>` is all you need to write.
    */

    // ------------------------------------------------------------
    // 12. MANUAL ARRAY-BASED STACK
    // ------------------------------------------------------------

    class ArrayStack {
        vector<int> data;

    public:
        void push(int x) {
            data.push_back(x);
        }

        void pop() {
            if (!data.empty())
                data.pop_back();
        }

        int top() const {
            if (data.empty())
                throw out_of_range("Stack is empty");

            return data.back();
        }

        bool empty() const {
            return data.empty();
        }

        int size() const {
            return (int)data.size();
        }
    };

    ArrayStack manualStack;

    manualStack.push(10);
    manualStack.push(20);

    cout << "Manual stack top: "
         << manualStack.top() << '\n';

    /*
        push       O(1) amortized
        pop        O(1)
        top        O(1)
        size       O(1)
    */

    // ------------------------------------------------------------
    // 13. MANUAL LINKED-LIST STACK
    // ------------------------------------------------------------

    class LinkedStack {
        struct Node {
            int value;
            Node* next;

            Node(int value, Node* next = nullptr)
                : value(value), next(next) {}
        };

        Node* head = nullptr;
        int n = 0;

    public:
        ~LinkedStack() {
            while (head)
                pop();
        }

        void push(int x) {
            head = new Node(x, head);
            ++n;
        }

        void pop() {
            if (!head)
                return;

            Node* temp = head;
            head = head->next;
            delete temp;
            --n;
        }

        int top() const {
            if (!head)
                throw out_of_range("Stack is empty");

            return head->value;
        }

        bool empty() const {
            return head == nullptr;
        }

        int size() const {
            return n;
        }
    };

    LinkedStack linked;

    linked.push(100);
    linked.push(200);

    cout << "Linked stack top: "
         << linked.top() << '\n';

    /*
        push  O(1)
        pop   O(1)
        top   O(1)
        size  O(1)
        Extra storage: O(n)
    */

    // ------------------------------------------------------------
    // 14. C++23: push_range
    // ------------------------------------------------------------

    /*
        In C++23, stack has push_range().

        Example:

        vector<int> values = {1,2,3};
        stack<int> st;
        st.push_range(values);

        This is an EXTRA feature.
        For DSA revision, push() is much more important.
    */

    // ------------------------------------------------------------
    // 15. NO DIRECT ITERATION
    // ------------------------------------------------------------

    /*
        std::stack does NOT expose:
            begin()
            end()
            operator[]
            front()
            back()

        You can only work through:
            top()
            push()
            pop()
            empty()
            size()

        To inspect all elements, copy the stack and pop from the copy.
    */

    return 0;
}

File: Library/DSA Reference/Data Structures/5) Stacks/stacks.py
"""
    ============================================================
    STACKS CHEAT SHEET — Python 3.x
    ============================================================

    CORE / MUST KNOW
    ----------------
    Primary DSA stack implementation: list

    Stack = LIFO
        Last In, First Out

    append / push       O(1) amortized
    pop last            O(1)
    top                 O(1) via a[-1]
    size                O(1)
    empty               O(1)

    Python also has collections.deque, which supports O(1)
    append/pop operations from either end and can be used as a
    stack.

    EXTRAS / REFERENCE
    ------------------
    deque, custom stack classes, linked-list stack, peek helpers,
    copying/printing, and implementation details.
"""

from collections import deque


# ================================================================
# CORE — MUST KNOW
# ================================================================

# ------------------------------------------------
# 1. INITIALISATION
# ------------------------------------------------

stack = []


# ------------------------------------------------
# 2. PUSH
# ------------------------------------------------

stack.append(32)
stack.append(10)
stack.append(20)

# Stack:
# top -> 20
#        10
#        32

# TC: O(1) amortized


# ------------------------------------------------
# 3. TOP / PEEK
# ------------------------------------------------

if stack:
    top = stack[-1]                 # O(1)
    print("Top:", top)

# IMPORTANT:
# stack[-1] on an empty list raises IndexError.


# ------------------------------------------------
# 4. POP
# ------------------------------------------------

if stack:
    x = stack.pop()                 # O(1)
    print("Popped:", x)

# Unlike C++ stack.pop(), Python list.pop()
# returns the removed element.


# ------------------------------------------------
# 5. SIZE
# ------------------------------------------------

size = len(stack)                   # O(1)
print("Size:", size)


# ------------------------------------------------
# 6. EMPTY
# ------------------------------------------------

is_empty = not stack                # O(1)
print("Empty:", is_empty)


# ------------------------------------------------
# 7. COMMON STACK PATTERN
# ------------------------------------------------

while stack:
    print(stack[-1], end=" ")
    stack.pop()

print()

# TC: O(n)
# SC: O(1) auxiliary
#
# This destroys the stack.


# ------------------------------------------------
# 8. NON-DESTRUCTIVE PRINTING
# ------------------------------------------------

original = [1, 2, 3, 4]

copy_stack = original.copy()        # O(n)

while copy_stack:
    print(copy_stack[-1], end=" ")
    copy_stack.pop()

print()

print("Original preserved:", original)

# Copying costs O(n) time + O(n) space.


# ================================================================
# EXTRAS — REFERENCE
# ================================================================

# ------------------------------------------------
# 9. DEQUE AS STACK
# ------------------------------------------------

dq_stack = deque()

dq_stack.append(10)
dq_stack.append(20)

top = dq_stack[-1]                  # O(1)
x = dq_stack.pop()                  # O(1)

print("Deque stack popped:", x)

# deque is also useful when a problem needs both-end operations.


# ------------------------------------------------
# 10. CUSTOM STACK CLASS
# ------------------------------------------------

class Stack:
    def __init__(self):
        self._data = []

    def push(self, x):
        self._data.append(x)

    def pop(self):
        if not self._data:
            raise IndexError("pop from empty stack")
        return self._data.pop()

    def top(self):
        if not self._data:
            raise IndexError("top from empty stack")
        return self._data[-1]

    def empty(self):
        return not self._data

    def size(self):
        return len(self._data)


s = Stack()

s.push(10)
s.push(20)

print("Custom stack top:", s.top())

# Complexity:
# push  O(1) amortized
# pop   O(1)
# top   O(1)
# size  O(1)


# ------------------------------------------------
# 11. MANUAL LINKED-LIST STACK
# ------------------------------------------------

class Node:
    def __init__(self, value, next_node=None):
        self.value = value
        self.next = next_node


class LinkedStack:
    def __init__(self):
        self.head = None
        self._size = 0

    def push(self, x):
        self.head = Node(x, self.head)
        self._size += 1

    def pop(self):
        if self.head is None:
            raise IndexError("pop from empty stack")

        value = self.head.value
        self.head = self.head.next
        self._size -= 1
        return value

    def top(self):
        if self.head is None:
            raise IndexError("top from empty stack")
        return self.head.value

    def empty(self):
        return self.head is None

    def size(self):
        return self._size


linked = LinkedStack()

linked.push(100)
linked.push(200)

print("Linked stack top:", linked.top())

# push  O(1)
# pop   O(1)
# top   O(1)
# size  O(1)
# Extra storage: O(n)


# ------------------------------------------------
# 12. STACK FROM TWO QUEUES / OTHER IMPLEMENTATIONS
# ------------------------------------------------

"""
A stack can also be implemented using other data structures,
for example two queues.

This is generally an interview / implementation exercise,
not the normal DSA implementation to use in code.

For practical Python DSA:
    list is usually enough.
"""


# ------------------------------------------------
# 13. COMMON STACK IDIOMS
# ------------------------------------------------

s = []

# Push:
s.append(x) if False else None

# Top:
top = s[-1] if s else None

# Safe pop:
item = s.pop() if s else None

# Empty:
if not s:
    pass

# Non-empty:
if s:
    pass


# ================================================================
# 14. CORE COMPLEXITY MEMORY
# ================================================================

"""
Python list stack:

push / append     O(1) amortized
pop last          O(1)
top               O(1)
size              O(1)
empty             O(1)

Avoid:
    pop(0)
    insert(0, x)

Those operate at the beginning of a list and cost O(n).
Use deque when you need efficient operations at both ends.
"""


# ------------------------------------------------
# 15. IMPORTANT PYTHON NOTES
# ------------------------------------------------

"""
1. Python does not have a dedicated built-in stack type.
2. list is the normal DSA stack implementation.
3. append() is push.
4. pop() is pop.
5. a[-1] is top/peek.
6. pop() returns the removed element.
7. pop() on an empty list raises IndexError.
8. a[-1] on an empty list raises IndexError.
9. pop(0) is O(n), so don't use it for a stack.
10. collections.deque is useful when both ends are needed.
"""
